codec/src/test/java/io/netty/handler/codec/ByteToMessageDecoderTest.java

```diff
@@ -562,4 +562,94 @@ protected void channelRead0(ChannelHandlerContext ctx, ByteBuf msg) throws Excep
         assertNotNull(channel.pipeline().get(FixedLengthFrameDecoder.class));
         assertFalse(channel.finish());
     }
+
+    @Test
+    public void testReuseInputBufferJustLargeEnoughToContainMessage_MergeCumulator() {
+        testReusedBuffer(Unpooled.buffer(16), false, ByteToMessageDecoder.MERGE_CUMULATOR);
+    }
+
+    @Test
+    public void testReuseInputBufferJustLargeEnoughToContainMessagePartiallyReceived2x_MergeCumulator() {
+        testReusedBuffer(Unpooled.buffer(16), true, ByteToMessageDecoder.MERGE_CUMULATOR);
+    }
+
+    @Test
+    public void testReuseInputBufferSufficientlyLargeToContainDuplicateMessage_MergeCumulator() {
+        testReusedBuffer(Unpooled.buffer(1024), false, ByteToMessageDecoder.MERGE_CUMULATOR);
+    }
+
+    @Test
+    public void testReuseInputBufferSufficientlyLargeToContainDuplicateMessagePartiallyReceived2x_MergeCumulator() {
+        testReusedBuffer(Unpooled.buffer(1024), true, ByteToMessageDecoder.MERGE_CUMULATOR);
+    }
+
+    @Test
+    public void testReuseInputBufferJustLargeEnoughToContainMessage_CompositeCumulator() {
+        testReusedBuffer(Unpooled.buffer(16), false, ByteToMessageDecoder.COMPOSITE_CUMULATOR);
+    }
+
+    @Test
+    public void testReuseInputBufferJustLargeEnoughToContainMessagePartiallyReceived2x_CompositeCumulator() {
+        testReusedBuffer(Unpooled.buffer(16), true, ByteToMessageDecoder.COMPOSITE_CUMULATOR);
+    }
+
+    @Test
+    public void testReuseInputBufferSufficientlyLargeToContainDuplicateMessage_CompositeCumulator() {
+        testReusedBuffer(Unpooled.buffer(1024), false, ByteToMessageDecoder.COMPOSITE_CUMULATOR);
+    }
+
+    @Test
+    public void testReuseInputBufferSufficientlyLargeToContainDuplicateMessagePartiallyReceived2x_CompositeCumulator() {
+        testReusedBuffer(Unpooled.buffer(1024), true, ByteToMessageDecoder.COMPOSITE_CUMULATOR);
+    }
+
+    private static void testReusedBuffer(ByteBuf buffer, boolean partial2, ByteToMessageDecoder.Cumulator cumulator) {
+        ByteToMessageDecoder decoder = new ByteToMessageDecoder() {
+            @Override
+            protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
```

Comment:
```suggestion
            protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) {
```

Reply:
I could make it more specific: `UnsupportedEncodingException`, since with the compatibility level of 1.6 I can't use `StandardCharsets.UTF_8`, as that wasn't introduced until Java 1.7, making the string `"UTF-8"` the most straightforward way of specifying the charset.

Reply:
Never mind. I found CharsetUtil.